lib/lte_link_control/lte_lc.c

```diff
@@ -632,20 +632,45 @@ static int connect_lte(bool blocking)
 	int err;
 	int tries = (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) ? 2 : 1);
 	enum lte_lc_func_mode current_func_mode;
+	enum lte_lc_nw_reg_status reg_status;
+	static atomic_t in_progress;
 
 	if (!is_initialized) {
 		LOG_ERR("The LTE link controller is not initialized");
 		return -EPERM;
 	}
 
+	err = lte_lc_nw_reg_status_get(&reg_status);
+	if (err) {
+		LOG_ERR("Failed to get current registration status");
+		return -EFAULT;
+	}
+
+	/* Do not attempt to register with an LTE network if the device already is registered.
+	 * This check is needed for blocking _connect() calls to avoid hanging for
+	 * CONFIG_LTE_NETWORK_TIMEOUT seconds waiting for a semaphore that will not be given.
+	 */
+	if ((reg_status == LTE_LC_NW_REG_REGISTERED_HOME) ||
+	    (reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING)) {
+		LOG_DBG("The device is already registered with an LTE network");
+
+		return 0;
+	}
+
+	if (atomic_get(&in_progress)) {
```

Comment:
Maybe move this check further up so that we can fail faster?

Reply:
The `in_progress` check is now before the registration status check